core/src/main/java/org/infinispan/persistence/spi/NonBlockingStore.java

```diff
@@ -0,0 +1,371 @@
+       * <p>
+       * Stores that have this characteristic must override the
+       * {@link #prepareWithModifications(Transaction, BatchModification)}, {@link #commit(Transaction)} and
+       * {@link #rollback(Transaction)} methods.
+       */
+      TRANSACTIONAL,
+      /**
+       * Whether this store supports being segmented. All methods in this SPI take as an argument a way to map a given
+       * entry to a segment. A segment in Infinispan is an int that acts as a bucket for many keys. Many store
+       * implementations may be able to store and load entries in a more performant way if they segment their data
+       * accordingly.
+       * <p>
+       * If this store is not segmentable then invokers of this SPI are not required to calculate these segments before
+       * invoking these methods and thus these methods may be invoked with any int value, null or equivalent. Please
+       * see each method to see how they may be affected when this store is not segmentable.
+       * <p>
+       * Note that a store may also be configured at runtime to be segmented or not. If this store is configured to not
+       * be segmented this store will be treated as if it does not have the SEGMENTABLE characteristic (causing possible
+       * parameters to be null or invalid segment numbers). A store implementation may want to block this configuration
+       * by throwing an exception in the {@link #start(InitializationContext)} method if it does not want to support this.
+       * <p>
+       * While it is possible that a SEGMENTABLE store can be configured as not segmented, a store that is not
+       * SEGMENTABLE will never be allowed to be configured as segmented.
+       * <p>
+       * Stores that have this characteristic must override the {@link #addSegments(IntSet)} and
+       * {@link #removeSegments(IntSet)} methods.
+       */
+      SEGMENTABLE,
+      /**
+       * If this store supports storing expiration metadata. Certain methods may or may not include expired entries.
+       * <p>
+       * Stores that have this characteristic must override the {@link #purgeExpired()} method.
+       */
+      EXPIRATION
+   }
+
+   /**
+    *
+    * @param ctx initialization context used to initialize this store
+    * @return a stage that when complete signals taht this store has been successfully started
+    */
+   CompletionStage<Void> start(InitializationContext ctx);
+
+   /**
+    *
+    * @return a stage that when complete signals that this store has been stopped
+    */
+   CompletionStage<Void> stop();
+
+   /**
+    *
+    * @return
+    */
+   default Set<Characteristic> characteristics() {
+      return EnumSet.noneOf(Characteristic.class);
+   }
+
+   /**
+    *
+    * @param storageMediaType
+    * @param supportedMediaTypes
+    * @return
+    */
+   default MediaType getKeyMediaType(MediaType storageMediaType, Set<MediaType> supportedMediaTypes) {
```

Comment:
Could you fill in the javadoc for this method?

Reply:
Sure, give me a few. This one is a new method and concept, so I agree it should be explained a bit better first.

Reply:
Added some docs to it.

Reply:
Could this be somehow simplified to
```default MediaType getKeyMediaType()``` where the store would choose what format it wants data in and it should spit out? 
Infinispan would then convert to/from this format whenever calling the store, and would throw an error if the format is not supported. 

I am missing what use case the media type negotiation is covering.

Reply:
The idea is mostly centered around RemoteStore that could at some point could allow for some type of media type negotiation. But if we don't think this will happen I can remove the parameters.

The store can always just return without checking the parameters as well. The arguments just give more context and options if the store wants them.

Reply:
> The store can always just return without checking the parameters as well. The arguments just give more context and options if the store wants them.

+1 I would rather have this as an option even if it is not used by the majority of stores. We can  document this in the Javadoc to avoid confusing implementors.